rs/rosetta-api/ledger_canister_blocks_synchronizer/src/blocks.rs

```diff
@@ -1403,89 +1391,165 @@ impl Blocks {
+        // We store all transactions for a single rosetta block in the values of this hashmap
+        let icp_transactions_per_rosetta_block = RefCell::new(HashMap::new());
+
+        // A list of all icp block indices that we need to convert to rosetta blocks
+        let block_indices = next_block_indices.first_block_index..=certified_tip_index;
+
+        // We fetch the first icp block to extract the timestamp, the first icp transaction and the first parent_hash of the first rosetta block created
+        let current_rosetta_block_index = RefCell::new(next_block_indices.rosetta_block_index);
         let Block {
             parent_hash,
             timestamp,
-            transaction,
-        } = Block::decode(self.get_hashed_block(&block_index)?.block)?;
+            ..
+        } = Block::decode(
+            self.get_hashed_block(&next_block_indices.first_block_index)?
+                .block,
+        )?;
 
-        let transaction_hash = transaction.hash();
-        let mut rosetta_block = RosettaBlock {
-            index: next_block_indices.rosetta_block_index,
-            parent_hash: parent_hash.map(|h| h.into_bytes()),
-            timestamp,
-            transactions: [(block_index, transaction)].into_iter().collect(),
-        };
-        let mut transactions_hashes = HashSet::new();
-        transactions_hashes.insert(transaction_hash);
-        for block_index in block_indices {
-            let block = self.get_hashed_block(&block_index)?;
-            let transaction = Block::decode(block.block)?.transaction;
-            let transaction_hash = transaction.hash();
-            if block.timestamp == rosetta_block.timestamp
-                && !transactions_hashes.contains(&transaction_hash)
-            {
-                rosetta_block.transactions.insert(block_index, transaction);
-            } else {
-                let next_rosetta_block_index = rosetta_block.index + 1;
-                let parent_hash = rosetta_block.hash();
-                num_rosetta_blocks_created += 1;
-                self.store_rosetta_block(rosetta_block)?;
-                rosetta_block = RosettaBlock {
-                    index: next_rosetta_block_index,
-                    parent_hash: Some(parent_hash),
-                    timestamp: block.timestamp,
-                    transactions: [(block_index, transaction)].into_iter().collect(),
+        let current_rosetta_block_parent_hash = RefCell::new(parent_hash);
+        let current_rosetta_block_timestamp = RefCell::new(timestamp);
+        let mut connection = self
+            .connection
+            .lock()
+            .map_err(|e| format!("Unable to aquire the connection mutex: {e:?}"))?;
+
+        let sql_tx = connection
+            .transaction()
+            .map_err(|e| format!("Unable to initialize a transaction: {e:?}"))?;
+
+        let mut statement1 = sql_tx
```

Comment:
Could these `statement1`, `statement2`, and `statement3` perhaps have some more descriptive names?